readme.txt
Grf2Html ReadMe


Contents:
 1 - About Grf2Html
 2 - Installation
 3 - Obtaining the source code
 4 - Usage
 5 - Errors: What to do, if Grf2Html prints error messages into the output?
 6 - Limitations
 7 - Notes


1) About Grf2Html:
   This tool converts TTDPatch [1] NewGrfs [2] into an easier readable html document.
   It should be capable of all(*) NewGrf features up to the date printed by the executable.

   The output is meant to debug your NewGrfs on a semantic level.
   Grf2Html is NOT meant to find syntactical errors of your NFO code. Use NfoRenum [3] instead.
   The generated HTML annotates several codes with their names. These texts are taken from the
   NewGraphicsSpecs [2] and from the source code documentation of TTDPatch [1].

   Grf2Html is published under the terms of GNU General Public License [4].

   The windows version of Grf2Html makes use of PNG Delphi by Gustavo Daud [5].

(*) See sections 6 and 7 about errors and limitations for details.


2) Installation:
   Download the binary build for your operating system from
      http://www.tt-forums.net/viewtopic.php?f=26&t=34279
   and extract it to your favorite location.
   About the usage proceed with section 4.


3) Compiling:
   If you do not want to use the above binaries but want to compile Grf2Html yourself, these are your steps.
   You can compile Grf2Html either on windows with Delphi or with Free Pascal on several operating systems,
   though compiling with Free Pascal is only tested on linux.

3.1) Compiling on windows with Delphi:
   You need:
     1) Delphi 5 or newer.
     2) A resource compiler. Usally "brcc32.exe" in your delphi\bin directory.
     3) A SubVersion (svn) client to get the source.
     4) PngDelphi by Gustavo Daud. Either from sourceforge [5] or from
           svn://dumbledore.zernebok.com/projects/3rdparty/pngdelphi
     5) If your binary should work on older windows versions (like 98) you need the fix for PngDelphi from
           http://sourceforge.net/tracker/index.php?func=detail&aid=1559286&group_id=16012&atid=116012
     6) Grf2Html source from
           svn://dumbledore.zernebok.com/projects/grf2html
   To build Grf2Html proceed in these steps:
     1) Apply the fix for PngDelphi if needed.
     2) Compile the resource files in the Grf2Html source directory. I.e. run from console:
           brcc32.exe grfbase.rc
           brcc32.exe tables.rc
        Typically the .rc extention is associated to brcc32 in explorer.
     3) Open "grf2html.dpr" with Delphi.
     4) Specify the location of PngDelphi. Depending on your Delphi version/language somewhere like
           Menu->Project->Options->Directories->SearchPath
     5) Press compile.
     6) Optionally run "grf2html --writeini" to create the default ini-file, that is bundled with the normal win32 build.

3.2) Compiling on linux with Free Pascal:
   You need:
     1) Free Pascal (fpc). Should be a package of your linux distribution.
     2) A resource compiler. Usally "windres", part of GNU binutils (though not always in the default build) or of mingw.
     3) A SubVersion (svn) client to get the source.
     4) libpng.
     5) Grf2Html source from
           svn://dumbledore.zernebok.com/projects/grf2html
   To build Grf2Html proceed in these steps:
     1) Adjust "Makefile" so it finds "windres". (In mingw it is named different)
     2) Run "make".
     3) Optionally run "./grf2html --writeini" to create the default ini-file, that is bundled with the normal binary.


4) Usage:
   Run Grf2Html from a console using:
      grf2html [options] <inputfiles ...>
   <inputfile ...>   Grfs to decode. Important: Specify an encoded .grf (not a .nfo).
   Options:
    -h                Prints this message and exits.
    --ini <file>      Reads default values from <file>. Default "grf2html.ini".
    --nodata          Skip generation of non-html data files (images, binary included data, ...).
    -o <path>         Output files to <path>/<file>/... If empty use path of the inputfile.
    -p <pal>          Specifies the palette to use in decoding: "win" or "dos".
    -r <first>:<last> Only generate output for a range of spritenumbers.
    -u <first>:<last> Only generate non-html data files in a range of sprites. Behaves like '--nodata' for sprites outside of the range.
    -v                Prints used options
    -w <width>        Aimed width for content frame in pixels. Used to determine number of columns in output.
    --writeini        Prints current options to the file specified by "-ini".

   However you can as well add Grf2Html to the explorer (or similiar program of your OS) context menu for files with .grf extention.
   In that case you want probably put your default options into "grf2html.ini", though they perhaps already suit your needs.
   Also note that not all options are available from the commandline. See section 5 for the options in the inifile.

   IMPORTANT:
   Depending on your operating- and file-system decoding a NewGrf can take some time. This is because Grf2Html outputs a lot files.
   For example a station NewGrf typically results in some 10000 files of hardly one kilobyte each. You will notice that simply copiing
   the result to a different location will take nearly the same amount of time as generating them in the first place.
   If you run Grf2Html with the '-nodata' switch, the generation of these files is skipped.
   The html output will still link to those files though. This is useful, if you have to decode your NewGrf multiple times with altered
   NFO code, but unchanged images; or if you are not interested in the images at all. Note that the bounding box previews of sprite
   layouts won't be generated either.


5) The inifile:
   This section describes the options in the inifile. Only the important options are also available on the commandline (see section 4).
   You find the inifile in the same directory as the executable (as long as you do not specify an other file with the "--ini <file>"
   command-line option. If the file does not exist yet, it is recommended to create it with the "--writeini" option to get all keys
   with their default values into the correct sections.

5.1) [Grf2Html] section:
   AimedWidth:              Works the same as the "-w" option.
   EntityFrame:             Enabled/Disabled the left entity frame. Specify "yes" or "no".
   IndexFrame:              Enables/Disabled the left index frame. Specify "yes" or "no".
   OutputPath:              Works the same as the "-o" option.
   Palette:                 Works the same as the "-p" option. Specify "win" or "dos".
   SubSpriteIndex:          Specifies if the left frame should also contain links to subsprites (e.g. realsprites in
                            Action1/5/A/12, sounds in Action11). Specify "yes" or "no".
   Transparency:            Controls whether transparent-blue of realsprites should be shown as "blue" or as "real" transparent.
   Verbose:                 Set to "1" or "0" to enable or disable verbose output ("-v" option) by default.

5.2) [Format] section:
   In this section are some magical values regarding cetain widths (in pixel) of the output. You may want to modify them, if you have
   larger/smaller fonts that me in your browser and the number-of-columns guessing of the "-w" option does not generate acceptable results.
      Action0ColWidth          Assumed width of one column in Action0s. Used the guess number of values per row.
      Action0FirstColWidth     Assumed width of left column (property name) in Action0s. Used the guess number of values per row.
      Action0SubIndexColWidth  Width of the column of e.g. number of spritelayout or customlayout in Action0s.
      Action1ColWidth          The assumed width of a columns in Action1s is the maximum of the sprite width and this value.
                               Used to guess number of sprites per row.
      Action5A12ColWidth       The same as Action1ColWidth but for Action5/A/12.
      LinkFrameWidth           Width of the left frame.


6) Errors: What to do, if Grf2Html prints error messages into the output?
   Sometimes Grf2Html prints a list of errors into the generated html output.
   In that case Grf2Html had problems understanding the input file.
   If you do not know the source of the errors proceed in the following steps:
     1) Check the NewGrf with NfoRenum [3]. If it prints the same errors it is most likely the NewGrf's fault.
        NfoRenum's error messages will probably be more descriptive.
     2) Check if your version of Grf2Html is up-to-date and capable of understanding the newest NewGrf features, that the NewGrf might use.
        If needed update your copy of Grf2Html.
     3) Else it is probably the fault of Grf2Html. Post bug reports to:
           http://www.tt-forums.net/viewtopic.php?f=26&t=34279
        Your bugreport should specify the following information:
           1) Which version of Grf2Html did you use.
           2) Which NewGrf causes the error. If the NewGrf is not public available at GrfCrawler, you have to attach the NewGrf to your post.
              If you do not want to post your file there, send it per private message.
           3) A description, what failed. E.g. sprite number, what is wrong, ...


6) Limitations:
    - BasicAction2s create hyperlinks to Action1 sets. They always refer to the last Action1.
      This might be wrong, if Action1s are skipped conditionally.
    - The meanings of some varaction2 variables change during some callbacks. Grf2Html always prints the regular meaning.
    - The NFO code is interpreted directly from the file. Changes by Action6s are not taken into account. All bad things may happen.
    - ActionB always uses the english built-in-messages.


7) Notes:
   [1] see http://www.ttdpatch.net
   [2] see http://wiki.ttdpatch.net/tiki-index.php?page=NewGraphicsSpecs
   [3] see http://users.tt-forums.net/dalestan/nforenum.
   [4] see file "COPYING.txt"
   [5] see http://pngdelphi.sourceforge.net


changelog.txt
Version 0.5.2 - 2010-05-02
 - Update to current NewGrfSpecs:
     Add variable 0x25.
     Add global property 0x12.
     Add house variables 47, 66 and 67.
     Add railtype properties and variables.
     Add airport tile properties and variables.
     Add action7/9 conditions 0x0D and 0x0E.
     Add lots of new languages to Action4.
     Add support for additional groundsprites in spritelayouts.
 - Remove: Old 'new airports'.
 - Change: Print usage on invalid command-line options.
 - Fix: Do not store binary included files using the unmodified filename from the grf. Instead make it valid and unique.
 - Fix: Off-by-one error in decoding of scStoreCompressed sprites. Though it worked in most cases :)

Version 0.5.1 - 2008-10-27
 - Update to match NewGrfSpecs as of 2008-10-25
     Add string codes 0x9A 0x06, 0x9A 0x07 and 0x9A 0x08.
     Add RandomAction2 type 0x84.
     Vehicle IDs of Action3 and 4 are extended bytes now.
     Add variables 0x22/0xA2, 0x23/0xA3 and 0x24/0xA4.
     Update variables 0x02/0x82 and 0x5F.
     Update train property 0x1A.
     Add road vehicle property 0x20.
     Add ship property 0x1B.
     Add aircraft property 0x1B.
     Add vehicle variable 0x49.
     Add bridge properties 0x00 and 0x13.
     Add bridge variables 0x40 and 0x41.
     Add house properties 0x21 and 0x22.
     Add house variable 0x65.
     Add global property 0x11.
     Add industry variables 0x45 (not new, but missing) and 0x46.
     Add object properties 0x0B, 0x0C, 0x0D, 0x0E, 0x0F and 0x10.
     Add object variables 0x40, 0x41, 0x42 and 0x43.
 - Fix/Feature: Make grf decoding more verbose when encountering errors. Also do not treat every error as fatal error. This allows decoding of certain 'invalid' grfs out there.
 - Fix: UTF-8 handling was wrong for custom ActionB messages. (glx)

Version 0.5 - 2008-03-25
 - Update to current NewGrfSpecs:
     Add PatchVariables 0x12 and 0x13.
     Add bridge properties 0x10, 0x11 and 0x12.
 - Preliminary support of the preliminary NewAirports specs.
 - Feature: Make links to Action1 sets and Action2s bidirectional.
 - Feature: Add a new frame ("entity"-frame) to group actions, which refer to the same ID of the same feature.
 - Feature: Add some new command line options:
     Allow setting an output-path.
     Allow restricting the output to a specific sprite range. (useful for big files)
     Generate data files (images, etc) only for a specific sprite range. (like "--nodata", but only inside a range)
 - Feature: Add lots of ini file options (in addition to the usual command line options):
     Enable/disable the two frames on the left.
     Enable/disable omitting of subsprites (e.g. images in Action1/5/A/12, sound data) from the left index-frame.
     Specify if transparent-blue should be displayed as blue or as "real" transparency.
     Ability to modify some magic contants used while generating the output, like pixel-widths of various html-elements in the output.
 - Feature: Improve readabilty of various actions:
     Print signed values of Action0 properties as signed numbers.
     Nicer bracket placement in AdvVarAct2.
     Compute var1A constants and insert them directly.
     Print range of used random bits in RandomAction2.
     Make ActionD print constants as signed/unsigned decimal numbers in some cases, instead of always hex.
     Place a space after the '#' of spritenumbers.
     Make background of spritelayout-previews transparent, instead of white.
 - Change: The command line parsing is now more tollerant towards spaces in some places.
 - Fix: Interpret string control characters 0x01, 0x0D, 0x0E, 0x0F and 0x1F.
 - Fix: Industry layout table row/column headers were wrong in some cases.
 - Fix: General variables in VarAction2 and in Action7/9/D are the same.

Version 0.4 - 2007-11-16
 - Port to Free Pascal to enable compiling for Linux. (Thanks TrueLight for helping)
 - Update to current NewGrfSpecs:
     Action0 IndustryProperties 23 and 24.
     Action5 Types 13, 14 and 15.
     Action5 with offset.
     Action7/9/D variable A1.
     ActionB message 06.
 - Feature: Allow passing of multiple files and wildcards by the command line.
 - Feature: Add support for TTDP string codes.
 - Feature: Handle characters in the range #$9E..#$FF correctly.
 - Feature: Print exceptions into the output and properly flush the files. (Belugas)
 - Feature: Allow to run the program even if (O)TTD(P) is running with the given grf file loaded (Belugas)
 - Change: Nicer output for BasicAction2 of Canals
 - Change: Make the operations in AdvancedVarAction2 and ActionD more readable by putting the "[signed]" etc. in a subscript.
 - Change: Nicer output for empty Action1/5/A/12 and empty Action0ByteArray (like HouseCargoAcceptanceWatchList or IndustryRandomSoundEffects)
 - Fix: Positions of bounding boxes in sprite layouts are signed. (George)
 - Fix: ActionB <data> field now handles UTF-8 correctly. (Though the <UTF-8> control code looks a bit missplaced in the output).
 - Fix: Creation of images crashed on older windows versions.

Version 0.3.1 - 2007-10-11
 - Fix: The first sprite is never a BinaryIncludeSprite. (boekabart)
 - Fix: Do not try to save binary included files with invalid filenames.
 - Fix: Test for end of Station Custom Layout was logically inverted. (boekabart)
 - Fix: Typos and mistakes in resources (PatchVars 99, 9A; GRM Mark operation).

Version 0.3 - 2007-10-04
 - Feature: UTF-8 support for texts.
 - Feature: Most 80+x variables of vehicles, stations, towns and industries are known. (Extracted from TTDP source documentation)
 - Fix: Text after </body> tag. (gmyx)
 - Fix: Decompression failed on some .grf. (minime)
 - Fix: Some spellings. (DaleStan)
 - Fix: Access violation, if industry layout did not start at (0,0). (belugas)
 - Fix: ActionB did not knew of "do not ignore during initialization".

Version 0.2 - 2007-10-02
 - First public release.
 - Support for Actions 0 to 13 as of current NewGraphicsSpecs.

newgrf.pas
(* This file is part of Grf2Html.
 * Copyright 2007 by Christoph Elsenhans.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *)
unit newgrf;

interface

uses sysutils, classes, contnrs, grfbase, nfobase, nfoact0, nfoact123, nfoact5A12, nfoact, outputsettings;

const
   grf2HtmlVersion                      : string = 'Grf2Html 0.4';
   dataVersion                          : string = '6th February 2008';

type
   TNewGrfFile = class
   private
      fGrfName     : string;
      fSprites     : TObjectList;
   public
      constructor create(aGrfName: string; grfFile: TObjectList);
      destructor destroy; override;
      procedure printHtml(path: string; settings: TGrf2HtmlSettings);
      property grfName: string read fGrfName write fGrfName;
      property sprites: TObjectList read fSprites;
   end;

procedure printAbout;

implementation

procedure printAbout;
begin
   writeln(grf2HtmlVersion);
   writeln('Copyright 2007-2008 by Christoph Elsenhans.');
   writeln;
   writeln('This program comes with ABSOLUTELY NO WARRANTY.');
   writeln('This is free software, and you are welcome to redistribute it');
   writeln('under certain conditions; see "COPYING.txt" for details.');
   writeln;
   writeln('Based on TTDPatch documentation and NewGraphicsSpecs from ', dataVersion); // they have written all the texts :)
   writeln('Visit http://wiki.ttdpatch.net/tiki-index.php?page=NewGraphicsSpecs');
   writeln;
   {$IFNDEF FPC}
      writeln('This tool makes use of PNG Delphi by Gustavo Daud'); // the author wants to get mentioned
      writeln('Visit http://pngdelphi.sourceforge.net');
      writeln;
   {$ENDIF}
end;

(*
 * Parses a Newgrf file.
 * @param aGrfName Filename of the grf (only used for printing it into the output)
 * @param grfFile  List of RealSprites, PseudoSprites and BinaryIncludeSprites generated by loadGrf (will be freed !).
 * @result         List of parsed NewgrfSprites, RealSprites, BinaryIncludeSprites and PseudoSprites (if unknown action).
 *)
constructor TNewGrfFile.create(aGrfName: string; grfFile: TObjectList);
var
   i, j, k                              : integer;

   msa                                  : TMultiSpriteAction;
   ssNr, ssCnt                          : integer;

   a1                                   : TAction1;
   action2Table                         : TAction2Table;
   a8                                   : TAction8;
   actionFTable                         : TActionFTable;
   action10Table                        : array[byte] of array of integer;

   src, dst                             : TSprite;
   psr                                  : TPseudoSpriteReader;
begin
   inherited create;
   fGrfName := aGrfName;
   grfFile.ownsObjects := false;
   fSprites := TObjectList.create(true);

   msa := nil;
   ssCnt := 0;
   ssNr := 0;

   a1 := nil;
   a8 := nil;
   for i := 0 to 255 do
   begin
      setLength(action10Table[i], 0);
      action2Table[i] := nil;
   end;
   for i := 0 to 127 do actionFTable[i] := nil;

   // Process NewGrf
   for i := 0 to grfFile.count - 1 do
   begin
      src := grfFile[i] as TSprite;
      if ssCnt > 0 then
      begin
         dst := msa.processSubSprite(ssNr, src);
         inc(ssNr);
         if ssNr = ssCnt then
         begin
            ssCnt := 0;
            ssNr := 0;
            msa := nil;
         end;
      end else
      if src is TPseudoSprite then
      begin
         psr := TPseudoSpriteReader.create(src as TPseudoSprite);
         if i = 0 then dst := TSpriteCountSprite.create(psr) else
         begin
            case psr.peekByte of
               $00: dst := TAction0.create(psr);
               $01: begin
                       a1 := TAction1.create(psr);
                       dst := a1;
                    end;
               $02: dst := TAction2.readAction2(psr, action2Table, a1);
               $03: dst := TAction3.create(psr, action2Table);
               $04: dst := TAction4.create(psr);
               $05: dst := TAction5.create(psr);
               $06: dst := TAction6.create(psr);
               $07: dst := TAction7.create(psr);
               $08: begin
                       dst := TAction8.create(psr);
                       if a8 = nil then a8 := dst as TAction8 else
                                        (dst as TNewGrfSprite).error('Multiple Action8 found. Using first one, ignoring this one.');
                    end;
               $09: dst := TAction9.create(psr);
               $0A: dst := TActionA.create(psr);
               $0B: dst := TActionB.create(psr);
               $0C: dst := TActionC.create(psr);
               $0D: dst := TActionD.create(psr);
               $0E: dst := TActionE.create(psr);
               $0F: dst := TActionF.create(psr, actionFTable);
               $10: begin
                       dst := TAction10.create(psr);
                       j := (dst as TAction10).labelNr;
                       k := length(action10Table[j]);
                       setLength(action10Table[j], k + 1);
                       action10Table[j][k] := i;
                    end;
               $11: dst := TAction11.create(psr);
               $12: dst := TAction12.create(psr);
               $13: dst := TAction13.create(psr);
               else dst := src;
            end;
         end;
         psr.free;
      end else dst := src;

      fSprites.add(dst);
      if src <> dst then src.free;

      if dst is TMultiSpriteAction then
      begin
         msa := dst as TMultiSpriteAction;
         ssCnt := msa.subSpriteCount;
         ssNr := 0;
      end;
   end;
   if ssCnt > 0 then msa.error('Unexpected end of file: ' + intToStr(ssCnt - ssNr) + ' more sprites expected.');

   // Set up Action7/9 destinations and Action8 links
   for i := 0 to fSprites.count - 1 do
   begin
      if fSprites[i] is TNewGrfSprite then (fSprites[i] as TNewGrfSprite).useAction8(a8);
      if fSprites[i] is TAction79 then
         with fSprites[i] as TAction79 do
         begin
            j := skip;
            if length(action10Table[j]) > 0 then
            begin
               // jump to label
               destination := fSprites[action10Table[j][0]] as TSprite;
               for k := 0 to length(action10Table[j]) - 1 do
               begin
                  if action10Table[j][k] > i then
                  begin
                     destination := fSprites[action10Table[j][k]] as TSprite;
                     break;
                  end;
               end;
            end else
            begin
               // skip sprites
               if (j > 0) and (i + j + 1 < fSprites.count) then destination := fSprites[i + 1 + j] as TSprite;
            end;
         end;
   end;
   grfFile.free;
end;

destructor TNewGrfFile.destroy;
begin
   fSprites.free;
   inherited destroy;
end;

(*
 * Generates a html file out of a Newgrf.
 * @param path         Path, where to store the output. With trailing backslash.
 * @param grfName      Filename of Grf, used for some titles.
 * @param newgrf       List of parsed Sprites.
 * @param aimedWidth   Approximated width for output in pixels. Used to guess number of columns in some tables.
 * @param suppressData If true, do not generate any files except the actual html files.
 *)
procedure TNewGrfFile.printHtml(path: string; settings: TGrf2HtmlSettings);
var
   t, t2                                : textFile;
   b, b2                                : array[word] of byte; // text file buffers
   i                                    : integer;
   s                                    : TSprite;
   ssCnt, spriteCount                   : integer;
begin
   if settings.range[0] < 0 then settings.range[0] := 0;
   if settings.range[1] >= sprites.count then settings.range[1] := sprites.count - 1;
   if settings.range[1] >= settings.range[0] then
   begin
      // Enlarge settings.range to include whole MultiSpriteActions
      for i := settings.range[1] downto 0 do
      begin
         if sprites[i] is TMultiSpriteAction then
         begin
            ssCnt := (sprites[i] as TMultiSpriteAction).subSpriteCount;
            if (i < settings.range[0]) and (i + ssCnt >= settings.range[0]) then settings.range[0] := i;
            if i + ssCnt > settings.range[1] then settings.range[1] := i + ssCnt;
         end;
      end;

      if settings.range[1] >= sprites.count then settings.range[1] := sprites.count - 1;
      spriteCount := settings.range[1] - settings.range[0] + 1;
   end else spriteCount := 0;

   assignFile(t, path + 'index.html');
   rewrite(t);
   writeln(t, '<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">');
   writeln(t, '<html><head>');
   writeln(t, '<title>' + fGrfName + '</title>');
   writeln(t, '<meta http-equiv="content-type" content="text/html; charset=utf-8">');
   writeln(t, '<meta name="generator" content="', grf2HtmlVersion, '">');
   writeln(t, '</head>');
   writeln(t, '<frameset cols="', settings.linkFrameWidth, ',*"><frame src="sprites.html" name="sprites" noresize><frame src="nfo.html" name="content">');
   writeln(t, '<noframes><body><a href="nfo.html">Content of grf</a></body></noframes>');
   writeln(t, '</frameset></html>');
   closeFile(t);

   assignFile(t, path + 'sprites.html');
   setTextBuf(t, b, sizeof(b));
   rewrite(t);
   writeln(t, '<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">');
   writeln(t, '<html><head>');
   writeln(t, '<title>Sprites in ', fGrfName, '</title>');
   writeln(t, '<meta http-equiv="content-type" content="text/html; charset=utf-8">');
   writeln(t, '<meta name="generator" content="', grf2HtmlVersion, '">');
   writeln(t, '</head><body><table summary="Sprite Index" width="100%">');

   assignFile(t2, path + 'nfo.html');
   setTextBuf(t2, b2, sizeof(b2));
   rewrite(t2);
   writeln(t2, '<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">');
   writeln(t2, '<html><head>');
   writeln(t2, '<title>', fGrfName, '</title>');
   writeln(t2, '<meta http-equiv="content-type" content="text/html; charset=utf-8">');
   writeln(t2, '<meta name="generator" content="', grf2HtmlVersion, '">');
   writeln(t2, '</head><body><table summary="Grf Content" width="100%" border="1" rules="rows">');

   ssCnt := 0;
   write('Generating html  0%');
   for i := settings.range[0] to settings.range[1] do
   begin
      if i * 100 div spriteCount <> (i - 1) * 100 div spriteCount then write(#13'Generating html', (i * 100 div spriteCount): 3, '%');

      s := fSprites[i] as TSprite;
      if (ssCnt = 0) or (settings.subSpritesInIndex = boolYes) then
      begin
         writeln(t, '<tr><td align=right>', i, '</td><td>', s.printHtmlSpriteLink('sprites', false), '</td></tr>');
      end;

      if ssCnt > 0 then dec(ssCnt) else
      begin
         writeln(t2, '<tr valign="top"><td align=right>', s.printHtmlSpriteAnchor, s.printHtmlSpriteNr, '</td><td>');
         try
            s.printHtml(t2, path, settings);
         except
            on E: Exception do
               begin
                  writeln('Exception: ', E.message);
                  writeln(t, '<tr><td colspan="2">Exception: ', E.message, '</td></tr>');
                  writeln(t2, '<br><br>Exception: ', E.message);
                  closeFile(t);
                  closeFile(t2);
                  halt;
               end;
         end;
         writeln(t2, '</td></tr>');
      end;

      if s is TMultiSpriteAction then ssCnt := (s as TMultiSpriteAction).subSpriteCount;
   end;

   writeln(#13'Generating html finished');
   writeln(t2, '</table>Generated by ', grf2HtmlVersion, '</body></html>');
   closeFile(t2);

   writeln(t, '</table></body></html>');
   closeFile(t);
end;

end.
